File: solutions/prolog/say/1/say.pl
% Dedicated to Shree DR.MDD

say(0, "zero") :- !.
say(1, "one") :- !.
say(2, "two") :- !.
say(3, "three") :- !.
say(4, "four") :- !.
say(5, "five") :- !.
say(6, "six") :- !.
say(7, "seven") :- !.
say(8, "eight") :- !.
say(9, "nine") :- !.
say(10, "ten") :- !.
say(11, "eleven") :- !.
say(12, "twelve") :- !.
say(13, "thirteen") :- !.
say(14, "fourteen") :- !.
say(15, "fifteen") :- !.
say(16, "sixteen") :- !.
say(17, "seventeen") :- !.
say(18, "eighteen") :- !.
say(19, "nineteen") :- !.

say(20, "twenty") :- !.
say(30, "thirty") :- !.
say(40, "forty") :- !.
say(50, "fifty") :- !.
say(60, "sixty") :- !.
say(70, "seventy") :- !.
say(80, "eighty") :- !.
say(90, "ninety") :- !.

say(Val, Output) :-
    between(10, 99, Val),
    Part is Val mod 10,
    Base is Val - Part,
    say(Base, TStr),
    say(Part, UStr),
    format(string(Output), "~s-~s", [TStr, UStr]), !.

say(Val, Output) :- spell_block(Val, 100, 999, "hundred", Output), !.
say(Val, Output) :- spell_block(Val, 1000, 999999, "thousand", Output), !.
say(Val, Output) :- spell_block(Val, 1000000, 999999999, "million", Output), !.
say(Val, Output) :- spell_block(Val, 1000000000, 999999999999, "billion", Output).

spell_units_only(U, 0, Label, OutStr) :-
    say(U, UStr),
    format(string(OutStr), "~s ~s", [UStr, Label]), !.
spell_units_only(U, Rem, Label, OutStr) :-
    say(U, UStr),
    say(Rem, RStr),
    format(string(OutStr), "~s ~s ~s", [UStr, Label, RStr]).

spell_block(Value, Min, Max, Label, Final) :-
    between(Min, Max, Value),
    divmod(Value, Min, Q, R),
    spell_units_only(Q, R, Label, Final).

File: solutions/ruby/say/1/say.rb
# Say a number in English (well, American actually)
class Say
  UNDER_TWENTY = %w(
    zero one two three four five six seven eight nine ten
    eleven twelve thirteen fourteen fifteen sixteen seventeen
    eighteen nineteen
  ).freeze

  TENS = %w(
    zero ten twenty thirty forty fifty sixty seventy eighty ninety
  ).freeze

  POWERS = {
    1_000_000_000 => 'billion',
    1_000_000 => 'million',
    1_000 => 'thousand',
    100 => 'hundred'
  }.freeze

  def initialize(number)
    raise ArgumentError, 'Negative number not allowed' unless number >= 0
    raise ArgumentError, 'Number is too large ' unless number < 1e12

    @number = number
  end

  def in_english
    Say.say(@number)
  end

  class << self
    # Say a number
    def say(number)
      if number < 20
        under_20(number)
      elsif number < 100
        under_100(number)
      else
        over_100(number)
      end
    end

    # Say a number under 20
    def under_20(number)
      UNDER_TWENTY[number]
    end

    # Say a number between 20 and 99
    def under_100(number)
      tens = TENS[number / 10]
      if number % 10 == 0
        # e.g. forty
        tens
      else
        # e.g. forty-two
        "#{tens}-#{say(number % 10)}"
      end
    end

    # Say a number 100 or over
    def over_100(number)
      power = POWERS.keys.select { |p| p <= number }.max
      name = POWERS[power]
      say_power = "#{say(number / power)} #{name}"
      remainder = number % power
      if remainder == 0
        # e.g. two hundred
        say_power
      else
        # e.g. two hundred twenty-three
        "#{say_power} #{say(remainder)}"
      end
    end
  end
end